### PR TITLE
New SpeedDial with mostly all issues solved + Example

### PR DESCRIPTION
**Pull request by prateekmedia**
Example:

Test apk: You can download it from the artifacts of latest [action](https://github.com/darioielardi/flutter_speed_dial/actions).

How to use (Add this to your pubspec.yaml):
```yaml
 flutter_speed_dial: 
    git:
      url: git://github.com/darioielardi/flutter_speed_dial.git
      ref: new
```
The new approach instead of using Stack, uses ColorFiltered which basically adds evenOdd filter to backgroundOverlay so that it looks like their is a hole in backgroundOverlay with shape of SpeedDial, the plus point with this approach is that we can easily track the movement of speedDial using compositeTransformFollower and layerLink, the negative point is that we have to add additional onTap to backgroundOverlay due to the fact that speedDial becomes unclickable as there is additional layer above it with backgrounOverlay.

The better approach is with clippath or custompainter, but the thing is that I currently don't know about tracking the movement using clipPath and CustomPainter as they require hardcoded values where you want to paint, so when displaying snackbar the speedDialChildren and the hole will not follow the speedDial fab. 

Reference Resources:
- https://www.flutterclutter.dev/flutter/tutorials/how-to-cut-a-hole-in-an-overlay/2020/510/
- https://stackoverflow.com/questions/56923884/cutting-a-hole-in-a-shape-with-flutter-canvas 

TODO :
- [ ] Needs Testing before putting out in pub.dev
- [x] ~~Shape of background Overlay not applying to SpeedDial~~. Fixed using ShapeDecoration for container
- [x] ~~Bug with background and activeBackground Color + same goes for foreground Color Property.~~
- [x] ~~Speed dial overlay clip doesnt resize based on device changes like orientation change.~~ Fixed using ColorFiltered, but searching for a better way if available.
- [x] ~~Make DialRoot Work.~~ 
- [x] ~~ButtonSize for main FAB is not real~~
- [x] ~~SpeedDial Children's are unclickable due to Positioned Widget.~~

### CODE REVIEW

File: lib/src/speed_dial_direction.dart [hunk 1]
```diff
@@ -0,0 +1,6 @@
+enum SpeedDialDirection { Up, Down, Left, Right }
+
+extension EnumExtension on SpeedDialDirection {
+  /// Get Value of The SpeedDialDirection Enum like Up, Down, etc. in String format
+  String get value => this.toString().split(".")[1];
```
Comment (shilangyu) [inline lib/src/speed_dial_direction.dart:5]: This functionality is already in flutter: https://api.flutter.dev/flutter/foundation/describeEnum.html

Comment (prateekmedia) [reply lib/src/speed_dial_direction.dart:5]: Yes, but I liked this way cause it doesn't require importing any library.

**Comment by brownAddict** [comment]
I have a speedDial. I am using this "new" unreleased branch that's why I am putting this here, not in issues.

`SpeedDial(
  key: _speedDialKey,
  ...
  children: [
    SpeedDialChild(
      key: _addInventoryKey,
    ),
    SpeedDialChild(),
]
`
if I add a key in any of the speedDialChild (_addInventoryKey in the above code), it throws this error when I click open the SpeedDial

`
════════ Exception caught by widgets library ═══════════════════════════════════
Multiple widgets used the same GlobalKey.
The key [LabeledGlobalKey<ScaffoldState>#bbfce] was used by 2 widgets[
    AnimatedChild-[LabeledGlobalKey<ScaffoldState>#bbfce]
        animation: AnimationController#0e206(▶ 0.000)➩Interval(0.0⋯1.0)➩Cubic(0.25, 0.10, 0.25, 1.00)➩Tween<double>(0.0 → 1.0)
        state: _AnimatedState#9aa1a
    FloatingActionButton-[LabeledGlobalKey<ScaffoldState>#bbfce]
        foregroundColor: Color(0xff000000)
        backgroundColor: Color(0xffffffff)
        elevation: 6.0
        dependencies: [_LocalizationsScope-[GlobalKey#63478], _InheritedTheme]
══════════════════════════════════════════════════════════════════
`

**Comment by prateekmedia** [comment]
> I have a speedDial. I am using this "new" unreleased branch that's why I am putting this here, not in issues.
> 
> `SpeedDial( key: _speedDialKey, ... children: [ SpeedDialChild( key: _addInventoryKey, ), SpeedDialChild(), ] `
> if I add a key in any of the speedDialChild (_addInventoryKey in the above code), it throws this error when I click open the SpeedDial
> 
> `════════ Exception caught by widgets library ═══════════════════════════════════ Multiple widgets used the same GlobalKey. The key [LabeledGlobalKey<ScaffoldState>#bbfce] was used by 2 widgets�[ AnimatedChild-[LabeledGlobalKey<ScaffoldState>#bbfce] animation: AnimationController#0e206(▶ 0.000)➩Interval(0.0⋯1.0)➩Cubic(0.25, 0.10, 0.25, 1.00)➩Tween<double>(0.0 → 1.0) state: _AnimatedState#9aa1a FloatingActionButton-[LabeledGlobalKey<ScaffoldState>#bbfce] foregroundColor: Color(0xff000000) backgroundColor: Color(0xffffffff) elevation: 6.0 dependencies: [_LocalizationsScope-[GlobalKey#63478], _InheritedTheme] ══════════════════════════════════════════════════════════════════`

Can you please test by checking if you are reusing keys or try by removing key parameter from speed dial, also are there any line numbers listed in the errors, if so then please also check and report for them, it might be your fault too.

**Comment by brownAddict** [comment]
> > I have a speedDial. I am using this "new" unreleased branch that's why I am putting this here, not in issues.
> > `SpeedDial( key: _speedDialKey, ... children: [ SpeedDialChild( key: _addInventoryKey, ), SpeedDialChild(), ] `
> > if I add a key in any of the speedDialChild (_addInventoryKey in the above code), it throws this error when I click open the SpeedDial
> > `════════ Exception caught by widgets library ═══════════════════════════════════ Multiple widgets used the same GlobalKey. The key [LabeledGlobalKey<ScaffoldState>#bbfce] was used by 2 widgets�[ AnimatedChild-[LabeledGlobalKey<ScaffoldState>#bbfce] animation: AnimationController#0e206(▶ 0.000)➩Interval(0.0⋯1.0)➩Cubic(0.25, 0.10, 0.25, 1.00)➩Tween<double>(0.0 → 1.0) state: _AnimatedState#9aa1a FloatingActionButton-[LabeledGlobalKey<ScaffoldState>#bbfce] foregroundColor: Color(0xff000000) backgroundColor: Color(0xffffffff) elevation: 6.0 dependencies: [_LocalizationsScope-[GlobalKey#63478], _InheritedTheme] ══════════════════════════════════════════════════════════════════`
> 
> Can you please test by checking if you are reusing keys or try by removing key parameter from speed dial, also are there any line numbers listed in the errors, if so then please also check and report for them, it might be your fault too.

I have attached 2 screenshots. In the first one, you can see the entire speeddial and you can see that i only used _addInventoryKey once. Even if I remove _speedDialKey, it will still break **as soon as I click open the toggle button.**

In the second screen shot, you can see the error message hopefully better.
You should be able to replicate this error by adding a key to speedDialChild and opening the toggle button (when you tap on it).

Another slight inconvenience. I am a flutter package (https://pub.dev/packages/tutorial_coach_mark) to display in app tutorial (that why I need keys). But the problem is that this new version opens on top the tutorial. If you go to that package link, you will see an example of how the package should work - then imagine if one of those widgets there is on top the tutorial. That means clicking on it interacts with the widget, not the tutorial. When closed, speedDial is below the tutorial, but when open, the speedDialChild is above the tutorial. Clicking on the speedDialChild calls the ontap of the speedDialChild, not the onTap on the tutorial. But this is just minor inconvenience. 
Thank you for all the good work you are doing.

Screenshot 2021-05-02 174931
Screenshot 2021-05-02 175059

**Comment by prateekmedia** [comment]
Ok will see that.

**Comment by prateekmedia** [comment]
@brownAddict Fixed key error in latest commit.

**Comment by brownAddict** [comment]
> @brownAddict Fixed key error in latest commit.

@prateekmedia Thank you. It works perfectly now.
For the slight inconvenience I mentioned earlier, any plans to include it in this branch?

**Pull request merged**